pandas/_libs/reduction.pyx

```diff
@@ -499,7 +499,9 @@ def apply_frame_axis0(object frame, object f, object names,
                 # `piece` might not have an index, could be e.g. an int
                 pass
 
-            if not is_scalar(piece):
+            if isinstance(piece, list):
```

Comment:
Do we want to special case lists specifically here? How about special casing only those times when `copy` is known to have the `deep` keyword for `piece.copy(deep="all")` and letting everything else use a `deepcopy`? (eg., something like `df.groupby("a").apply(lambda x: set(x.index))` would still raise here)

Reply:
>  How about special casing only those times when copy is known to have the deep keyword

How - by using `inspect`?

Reply:
I had been thinking checking specifically for Series, DataFrame, or Index and trying to `deepcopy` everything else, but there may be a "prettier" way of approaching it. I remember getting the exact same test failure you're seeing now and am still puzzled by it.